### PR TITLE
throw when end distance not bigger than start distance in COMPSEGS

### PR DESCRIPTION
**Pull request by GitPaean**

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
Maybe this could be handled by the `ParseContext` mechanism?

**Comment by GitPaean** [comment]
> Maybe this could be handled by the `ParseContext` mechanism?

Excuse me if I did not follow this very closely. 

One question, will `handleError` form `ParseContext` terminate the running or it depends?

For the case in this PR, the correct thing is to terminate the running.

**Comment by GitPaean** [comment]
I see how it works now. Will update this PR with `ParseContext` later.

**Comment by GitPaean** [comment]
Mostly the update will be after the Easter. It is something correct to do, while it is just pointing out the deck should be fixed (we have such a deck, and it is to be fixed). It is not urgently needed if the deck is not wrong.

**Comment by GitPaean** [comment]
I am revisiting this PR and trying to finalizing it. 

I am a little struggling to find a proper Key for the situation in this PR. Should I introduce a new one?

Please suggest.  Basically this input is not sensible physically and the simulation will terminate since it will cause problem for running.

**Comment by joakim-hove** [comment]
> Should I introduce a new one?

Yes - `ParseContext` keys should be introduced liberally. I have tried to give the keys a sensible prefix - so maybe `SCHEDULE_INVALID_COMPSEGS` could be a good choice.

Since this is a serious problem - it should default to throw an exception.

**Comment by GitPaean** [comment]
It throws at the places it should be, while it throws twice. 
```
Failed to create valid EclipseState object.
Exception caught: SCHEDULE_INVALID_COMPSEGS:  The end of the perforations need be to further down than the start of the perforations
  well F 5 7 4 in keyword COMPSEGS

terminate called after throwing an instance of 'std::invalid_argument'
  what():  SCHEDULE_INVALID_COMPSEGS:  The end of the perforations need be to further down than the start of the perforations
  well F 5 7 4 in keyword COMPSEGS
```

**Comment by GitPaean** [comment]
It looks like it got caught twice and the second time the simulation is terminated.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
I am done updating. Just let jenkins test it.

**Comment by GitPaean** [comment]
The two places it throws are 
https://github.com/OPM/opm-common/blob/464bc4b795fce07f55903cf4a4f5a051c058993e/src/opm/parser/eclipse/Parser/ParseContext.cpp#L155

and 
https://github.com/OPM/opm-simulators/blob/a618a1a7771d0dd17a26bbc451a7bec2dfd5cdb7/flow/flow.cpp#L293

**Comment by GitPaean** [comment]
I could not reproduce it locally. and there is no more information to look at. I will perform the test again. 

```
      Start 37: compareECLFiles_flow+3D_MSW
37/57 Test #37: compareECLFiles_flow+3D_MSW .........................***Exception: Other  2.69 sec
Build step 'Execute shell' marked build as failure
[WARNINGS] Skipping publisher since build result is FAILURE
Archiving artifacts
Recording test results
```

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
> It looks like it got caught twice and the second time the simulation is terminated.

Yes - it throws twice; the second throw seems a bit weird to me - that should in my opinion just be `std::exit(1)`?!

File: tests/parser/MultisegmentWellTests.cpp [hunk 1]
```diff
@@ -121,3 +124,56 @@ BOOST_AUTO_TEST_CASE(MultisegmentWellTest) {
...
+    BOOST_CHECK_EQUAL( 8U, compsegs.size() );
+
+    Opm::WellSegments segment_set;
+    const Opm::DeckKeyword welsegs = deck.getKeyword("WELSEGS");
+    segment_set.loadWELSEGS(welsegs);
+
+    BOOST_CHECK_EQUAL(6U, segment_set.size());
+
+    Opm::ErrorGuard   errorGuard;
+    Opm::ParseContext parseContext;
```
Comment (joakim-hove) [inline tests/parser/MultisegmentWellTests.cpp]: In situations like this - where you actually depend on the `ParseContext` setting, I prefer to set it explicitly:
```
parseContext.update(ParseContext::SCHEDULE_INVALID_COMPSEGS, InputError::THROW_EXCEPTION);
BOOST_CHECK_THROW( );
```
if at all possible to proceed with an invalid `COMPSEGS` you could then add the additional test:
```
parseContext.update(ParseContext::SCHEDULE_INVALID_COMPSEGS, InputError::IGNORE);
BOOST_CHECK_NO_THROW( Opm::newConnectionsWithSegments( ... ) );
```

**Comment by joakim-hove** [comment]
I have tried to have the `ParseContext` keys "somewhat hierarchical" - i.e. I would prefer if you named them `SCHEDULE_COMPSEGS_INVALID` and `SCHEDULE_COMPSEGS_NOT_SUPPORTED`.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
```
Process leaked file descriptors. See https://jenkins.io/redirect/troubleshooting/process-leaked-file-descriptors for more information
Build step 'Execute shell' marked build as failure
[WARNINGS] Skipping publisher since build result is FAILURE
Archiving artifacts
```

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by akva2** [comment]
i'm sorry, i just cannot find the reason. i'm monitoring logs but i don't see anything particular happening. my googling fails, ...

**Comment by GitPaean** [comment]
> i'm sorry, i just cannot find the reason. i'm monitoring logs but i don't see anything particular happening. my googling fails, ...

Thanks for the efforts.  I am just reporting information and pointing out the failure is not necessarily because there is something wrong with the pull request.

**Comment by joakim-hove** [comment]
@GitPaean : In light of the current Jenkins difficulties I am happy to merge this on your word that it *works for you*.

**Comment by GitPaean** [comment]
> @GitPaean : In light of the current Jenkins difficulties I am happy to merge this on your word that it works for you.

Locally, it works fine.  Jenkins might go through this time, let us wait for it to finish.  You can go ahead to merge it no matter what the result is.

**Comment by GitPaean** [comment]
It is green. :)

**Pull request merged**